independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanGenerator.java

```diff
@@ -839,16 +849,21 @@ protected void implementDestroy(BeanInfo bean, ClassCreator beanCreator, Provide
                                     callback.declaringClass().name(), callback.name()));
                         }
                         reflectionRegistration.registerMethod(callback);
-                        destroy.invokeStaticMethod(MethodDescriptors.REFLECTIONS_INVOKE_METHOD,
-                                destroy.loadClass(callback.declaringClass().name().toString()),
-                                destroy.load(callback.name()), destroy.newArray(Class.class, destroy.load(0)),
-                                destroy.getMethodParam(0),
-                                destroy.newArray(Object.class, destroy.load(0)));
+                        preDestroyBytecode.invokeStaticMethod(MethodDescriptors.REFLECTIONS_INVOKE_METHOD,
+                                preDestroyBytecode.loadClass(callback.declaringClass().name().toString()),
+                                preDestroyBytecode.load(callback.name()),
+                                preDestroyBytecode.newArray(Class.class, preDestroyBytecode.load(0)),
+                                instance,
+                                preDestroyBytecode.newArray(Object.class, preDestroyBytecode.load(0)));
                     } else {
                         // instance.superCoolDestroyCallback()
-                        destroy.invokeVirtualMethod(MethodDescriptor.of(callback), destroy.getMethodParam(0));
+                        preDestroyBytecode.invokeVirtualMethod(MethodDescriptor.of(callback), instance);
                     }
                 }
+                if (preDestroyBytecode != destroy) {
```

Comment:
Nice trick but it could be useful to add some pseudocode in the comments to illustrate the difference...?

Reply:
I thought the comment below this line would be enough, but sure, I'll add something.

Reply:
Well, it's a _"go figure out yourself"_-style comment :D

Reply:
Done.